src/coreclr/jit/inductionvariableopts.cpp

```diff
@@ -0,0 +1,1533 @@
+        }
+        case GT_STORE_LCL_VAR:
+        {
+            GenTreeLclVarCommon* store = tree->AsLclVarCommon();
+            GenTree*             data  = store->Data();
+            if (!data->OperIs(GT_PHI))
+            {
+                return Analyze(block, data, depth + 1);
+            }
+
+            if (block != m_loop->GetHeader())
+            {
+                return nullptr;
+            }
+
+            // We have a phi def for the current loop. Look for a primary
+            // induction variable.
+            GenTreePhi*    phi         = data->AsPhi();
+            GenTreePhiArg* enterSsa    = nullptr;
+            GenTreePhiArg* backedgeSsa = nullptr;
+
+            for (GenTreePhi::Use& use : phi->Uses())
+            {
+                GenTreePhiArg*  phiArg = use.GetNode()->AsPhiArg();
+                GenTreePhiArg*& ssaArg = m_loop->ContainsBlock(phiArg->gtPredBB) ? backedgeSsa : enterSsa;
+                if ((ssaArg == nullptr) || (ssaArg->GetSsaNum() == phiArg->GetSsaNum()))
+                {
+                    ssaArg = phiArg;
+                }
+                else
+                {
+                    return nullptr;
+                }
+            }
+
+            if ((enterSsa == nullptr) || (backedgeSsa == nullptr))
+            {
+                return nullptr;
+            }
+
+            ScevLocal* enterScev = NewLocal(enterSsa->GetLclNum(), enterSsa->GetSsaNum());
+
+            LclVarDsc*    dsc    = m_comp->lvaGetDesc(store);
+            LclSsaVarDsc* ssaDsc = dsc->GetPerSsaData(backedgeSsa->GetSsaNum());
+
+            if (ssaDsc->GetDefNode() == nullptr)
+            {
+                // GT_CALL retbuf def
+                return nullptr;
+            }
+
+            if (ssaDsc->GetDefNode()->GetLclNum() != store->GetLclNum())
+            {
+                assert(dsc->lvIsStructField && ssaDsc->GetDefNode()->GetLclNum() == dsc->lvParentLcl);
+                return nullptr;
+            }
+
+            assert(ssaDsc->GetBlock() != nullptr);
+
+            // We currently do not handle complicated addrecs. We can do this
+            // by inserting a symbolic node in the cache and analyzing while it
+            // is part of the cache. It would allow us to model
+            //
+            //   int i = 0;
+            //   while (i < n)
+            //   {
+            //     int j = i + 1;
+            //     ...
+            //     i = j;
+            //   }
+            // => <L, 0, 1>
+            //
+            // and chains of recurrences, such as
+            //
+            //   int i = 0;
+            //   int j = 0;
+            //   while (i < n)
+            //   {
+            //     j++;
+            //     i += j;
+            //   }
+            // => <L, 0, <L, 1, 1>>
+            //
+            // The main issue is that it requires cache invalidation afterwards
+            // and turning the recursive result into an addrec.
```

Comment:
Any idea how prevalent these more complicated IVs might be?

Reply:
I haven't checked, but I have some (hacky) code lying around for it that I was planning to look at the impact of later.

There are some other notable missing things in the SCEV IR, like a subtraction operator, so loops that subtract by an invariant local are not representable (subtractions with a constant are usually represented as `GT_ADD` so those are ok).